direct_consumer/data_queue: fairness fix & tests

joe-redpanda opened the pull request:
Data queue is intended to bound the memory and number of elements that
can persist in its underlying queue by blocking callers when full.

Data queue does not block fairly though. A large push may be blocked
while small pushes are allowed to bypass the waiter queue and join the
data queue immediately.

In the case of many small pushes 'cutting the line' the waiter queue
will be starved indefinitely.

This pr
1. adds a regression tests which demonstrates queue unfairness in the prior implementation
2. a mask ontop of seastar::condition_variable which disallows the problematic conditional waits
3. the fix to make the queue fair
4. test upgrades
5. a chaos test to catch any other concurrency issue

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

mmaslankaprv commented:
Can you describe in the commit message what the problem was and  why the fix helps. Also consider moving the test into separate commit

joe-redpanda committed:
data_queue_test: add a fairness test

Data queue is intended to bound the memory and number of elements that
can persist in its underlying queue by blocking callers when full.

Data queue does not block fairly though. A large push may be blocked
while small pushes are allowed to bypass the waiter queue and join the
data queue immediately.

In the case of many small pushes 'cutting the line' the waiter queue
will be starved indefinitely.

This commit adds a test that demostrates this starving behavior. This
commit does NOT pass tests as a result.

joe-redpanda committed:
data_queue: fair multifiber push

This commit makes data_queue fair to multifiber pushes.

Why was the prior queue unfair?
1. small fetches are allowed to skip the waiter queue, this can block a
   waiting fetch forever
2. cv with conditional wait has issues
   - if you don't recheck the wait condition, the condition may be false
     by the time the waiter has awoken
   - if you do recheck the wait condition, the only reasonable option is
     to rejoin the waiter queue. This means a potentially infinite
     carosel of waits
   - the answer is don't do conditional waits
     - we already know all of the wait conditions, dont signal the cv if
       the next waiter would reevaluate to 'wait again' this lets the
       head of the queue remain unordered

How does this commit address this?
1. operations now check their respective waiter queues for waiting
   operations before invoking shortcut logic
   - example: even if the queue has more data, if theres currently
     waiting pops in the queue, this pop should wait in line
2. removing predicated waits on ss::condition_variable
   - we know the wait conditions before waking the next waiter, we
     simply don't wake the next waiter if its condition would return it
     to the waiter queue

Without conditional waits, pops and pushes need their own cv's, so
_updated splits into _waiting_for_pop and _waiting for push.

A push/pop may need to wake the next operation in sequence as well, so
conditional signals are added.

The behavior of data_queue is slightly modified. The limits on
queue count stay the same, but data_queue will accept pushes
until and including the first push that puts it oversize.

The tests have been updated accordingly to test the above.
The fairness test now passes

Added a chaos test to catch any other concurrency issues

mmaslankaprv commented on src/v/kafka/client/direct_consumer/data_queue_cv.h:
```diff
@@ -0,0 +1,37 @@
+namespace kafka::client::detail {
+
+// quick chop of condition variable to be better used as a fair waiter queue
+class data_queue_cv : private seastar::condition_variable {
```
why is this required ? It looks to me that the base condition has all necessary methods

joe-redpanda replied:
Not required, just discourages use of conditional wait, could remove

dotnwat commented on src/v/kafka/client/direct_consumer/data_queue_cv.h:
```diff
@@ -0,0 +1,37 @@
+namespace kafka::client::detail {
+
+// quick chop of condition variable to be better used as a fair waiter queue
+class data_queue_cv : private seastar::condition_variable {
```
>Adds data_queue_cv which is a mask on seastar::condition_variable which
hides conditional waits

instead of masking out interfaces that aren't desired, could the user of the data_queue_cv make its CV a private member and only call the methods it needs?

dotnwat commented on src/v/kafka/client/direct_consumer/data_queue.h:
```diff
@@ -75,11 +87,33 @@ class data_queue {
+    }
+
+    // wakes the next pop if appropriate
+    inline void update_next_pop() {
```
`inline` is implied, but perhaps move the impl of this method into the `.cc` file?

dotnwat commented on src/v/kafka/client/direct_consumer/data_queue.cc:
```diff
@@ -10,60 +10,91 @@
 namespace kafka::client {
 
+namespace {
+ss::logger log{"data_queue"};
```
i'm wondering if this logger is too fine grained? perhaps you want to take a reference to an existing logger?

dotnwat commented on src/v/kafka/client/direct_consumer/data_queue.cc:
```diff
@@ -10,60 +10,91 @@
  */
```
> This commit makes data_queue fair to multifiber pushes.

Why is it unfair?

joe-redpanda replied:
Starvation: fairness test demonstrates.
1. small fetches were allowed to skip the waiter queue which would block out a waiting fetch forever
2. cv with conditional wait has issues
    - If you don't recheck the wait condition, the wait condition can be false by the time the waiter is awoken
    - If you do recheck the wait condition, the best option is to rejoin the waiter queue which means a potentially infinite carousel of waits
    - The answer is just don't do conditional waits.
        - we already know all of the wait conditions, just dont signal the cv if the condition is 'wait again' so the queue head can remain the queue head

dotnwat replied:
that should be in the commit message

joe-redpanda replied:
updated, text of the 'fixing' commit pasted below:

```
This commit makes data_queue fair to multifiber pushes.

Why was the prior queue unfair?
1. small fetches are allowed to skip the waiter queue, this can block a
   waiting fetch forever
2. cv with conditional wait has issues
   - if you don't recheck the wait condition, the condition may be false
     by the time the waiter has awoken
   - if you do recheck the wait condition, the only reasonable option is
     to rejoin the waiter queue. This means a potentially infinite
     carosel of waits
   - the answer is don't do conditional waits
     - we already know all of the wait conditions, dont signal the cv if
       the next waiter would reevaluate to 'wait again' this lets the
       head of the queue remain unordered

How does this commit address this?
1. operations now check their respective waiter queues for waiting
   operations before invoking shortcut logic
   - example: even if the queue has more data, if theres currently
     waiting pops in the queue, this pop should wait in line
2. removing predicated waits on ss::condition_variable
   - we know the wait conditions before waking the next waiter, we
     simply don't wake the next waiter if its condition would return it
     to the waiter queue

Without conditional waits, pops and pushes need their own cv's, so
_updated splits into _waiting_for_pop and _waiting for push.

A push/pop may need to wake the next operation in sequence as well, so
conditional signals are added.

The behavior of data_queue is slightly modified. The limits on
queue count stay the same, but data_queue will accept pushes
until and including the first push that puts it oversize.

The tests have been updated accordingly to test the above.
The fairness test now passes

Added a chaos test to catch any other concurrency issues
```

joe-redpanda commented:
I'll fix the merge conflicts in a smidge, quick comment fixes and fixed an accidental use-after-move in the chaos test which  busted the test

dotnwat commented on src/v/kafka/client/direct_consumer/data_queue.cc:
```diff
@@ -10,60 +10,114 @@
 
 ss::future<fetches> data_queue::pop(std::chrono::milliseconds timeout) {
-    co_await _updated.wait(timeout, [this] { return !_queue.empty(); });
+    if (_queue.size() == 0 || _waiting_to_pop.has_waiters()) {
+        co_await _waiting_to_pop.wait(timeout);
+    }
+    vassert(
+      _queue.size() > 0,
```
i see that the _waiting_to_pop is only signaled if the queue is non-empty, so presumably that is the behavior that supports this assertion? but while that is true, what i can't figure out is if the assertion is guaranteed to pass. what happens if a fiber (called X) is scheduled between that signal and this fiber waking up and fiber X interacts with the data queue to invalidate that assertion?

joe-redpanda replied:
this is guarded by the conditional above
if there are waiters in the _waiting_to_pop queue, any new calls to pop are put on the back of the waiter queue.
Its a hard guarantee that the next pop that gets a queue element will be the queue head

joe-redpanda replied:
I'm going to add a quick check though that timed out waiters dont consume a signal. I would have hoped they didn't

joe-redpanda replied:
Done and success. CV behaves nicely with timeouts

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--71a52d82-b814-4b15-8151-a2fab001b9d5-->
<!--TEST RESULTS-->
<!--ba9e0faf-0a87-45ee-8e5d-0ec348909917-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70676">build#70676</a></summary>


| test_class            | test_method                   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------|:------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| FeaturesMultiNodeTest | test_license_upload_and_query | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70676#0198a026-12fc-4ddc-a281-b4a21c80fe4f | FLAKY         | 20/21    | upstream reliability is '91.93181818181819'. current run reliability is '95.23809523809523'. drift is -3.30628 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70817">build#70817</a></summary>


| test_class                                               | test_method                                              | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:---------------------------------------------------------|:---------------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| src/v/kafka/client/direct_consumer/tests/data_queue_test | src/v/kafka/client/direct_consumer/tests/data_queue_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/70817#0198a9a3-8ff8-4359-94ad-fe7a4433efb2 | FAIL          | 0/1      |          |
</details>
<!--/body-->

joe-redpanda commented:
Bug still persists but the odds of its causing a real problem are really really low.

Closing this in favor of a more generic solution I'll come back to.

Really what we wanted here was a queue that could be bounded on multiple criteria
This is not feasible with a semaphore without a hack
The logic here is over complicated for a single use queue

soution:
```
push_payload -> bound info for a push
queue_stm -> decides whether a push/pop should proceed / wait, decides whether a waiter should be woken
queue -> 
    std::queue<promise> -> pops which are waiting
    std::queue<pair<promise, push_payload>> -> pushes which are waiting + payload metadata s.t. we can peek at the front of the queue
```
The rest is essentially the same logic + some template specialization
should work generically for all your bounded queue needs + be easier to test & review
